RavenDB-6361 Review only - not ready to merge yet

fitzchak opened the pull request.

Commit:
RavenDB-6322 Implement stream get, without metadata.

Commit:
RavenDB-6361 We will use attachments instead of file system. Implement Get, Put and Delete.

Commit:
RavenDB-6361 Use 1 field to store the attachment key instead of 3. This way we can get it using GetPrimaryKey method, and we can also iterate using this key by call SeekBackwardFrom.
Fix bugs and code to another method: UpdateDocumentAfterAttachmentPut
Call AddAfterCommitNotification after we inserted the stream.

Commit:
RavenDB-6361 Allow content type to be null

Commit:
RavenDB-6361 Refactor SetDocumentChangeVectorForLocalChange method and avoid loading the document from the table twice

Commit:
RavenDB-6361 Fix data corruption. The Key on the document cannot be used for saving the key back to storage. RawKey is introduced in order to save other the future from this trap as it make it easy to see what key should be used for persistent.

Commit:
RavenDB-6361 Return the attachment names in the metadata as joined string with a comma: @attachments: "name1.txt, name2.png, name3.jpg".
Change UpdateDocumentAfterAttachmentPut to be more efficiant and avoid extra allocation in TableValueToDocument - remove RawKey.
Use a more user friendly stream identifier - regular etag instead of big endian.
Refactor code that is error prone (which also has an attention comment: See format of the lazy string key in the GetLowerKeySliceAndStorageKey method)

Commit:
RavenDB-6361 Put and Get documents / attachments commands should send the etag if it is not null

Commit:
RavenDB-6361 Get attachment using the API. WIP - only for review.

Commit:
RavenDB-6361 Fix bug

ayende commented on src/Raven.Client/Documents/Commands/GetDocumentCommand.cs:
```diff
@@ -83,6 +85,9 @@ public override HttpRequestMessage CreateRequest(ServerNode node, out string url
             };
 
+            if (Etag.HasValue)
+                request.Headers.IfNoneMatch.Add(new EntityTagHeaderValue($"\"{Etag.Value}\""));
```
Why is this here? This isn't the responsability of the command, it is handled by the command executer

fitzchak replied:
Indeed. Removed.

ayende commented on src/Raven.Client/Documents/Commands/GetDocumentCommand.cs:
```diff
@@ -18,6 +19,7 @@ public class GetDocumentCommand : RavenCommand<GetDocumentResult>
 
         public string[] Ids;
         public string[] Includes;
+        public long? Etag;
```
This shouldn't be here.

fitzchak replied:
Removed, as Get requests has the etag set by the request executer.

ayende commented on src/Raven.Client/Documents/Commands/PutDocumentCommand.cs:
```diff
@@ -29,6 +30,9 @@ public override HttpRequestMessage CreateRequest(ServerNode node, out string url
 
+            if (Etag.HasValue)
+                request.Headers.IfMatch.Add(new EntityTagHeaderValue($"\"{Etag.Value}\""));
+
```
Since we are making a modification here, we _do_ want it here (the executer won't do that for us for writes).
However, is the a way to avoid the string format / concat here?

fitzchak replied:
I modified it to be:
request.Headers.TryAddWithoutValidation("If-Match", $"\"{etag.Value}\"");
This will avoid no needed class allocation and some validation code but I don't see a way to avoid string concat as we need to pass a string there.
I thought that we do not care about this in the client side as it is going to make a request to the server.

ayende commented on src/Raven.Client/Documents/Operations/GetAttachmentOperation.cs:
```diff
@@ -0,0 +1,84 @@
+                    Method = HttpMethods.Get,
+                };
+
+                if (_etag.HasValue)
```
Why is this here? This is a read request, after all, and the executer should handle it.
That said, however, I'm not sure what it even should do for attachments, they are likely to be large, no?
So do we want to have them cached in the first place?

fitzchak replied:
No, we do not want to cache them in the client. So we do need to add it here.

fitzchak replied:
Actually as we are not going to cache this I'm going to remove this.

ayende replied:
We don't cache it, but the user's might, and it is a good idea to enable this scenario for them.
I'm fine with marking this as a non read request to disable the request executer caching, and then letting the user provide an etag manually.

ayende replied:
Note that we need to be clear about caching attachments in the docs, though

ayende commented on src/Raven.Client/Http/RavenCommand.cs:
```diff
@@ -62,16 +67,24 @@ public bool IsFailedWithNode(ServerNode leaderNode)
+                // ReSharper disable once PossibleInvalidOperationException
+                var etag = response.GetEtagHeader().Value;
+                // TODO: How do we make sure to keep the stream open?
+                // Should we write to stream that the user give us instead of return a stream?
```
Yes, it is probably better to have the user provide a destination stream that we'll write to immediately. 
This way they can choose how to handle this, directly to network, to file or process inline.

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -73,6 +79,28 @@ public enum ConflictsTable
+            StreamIdentifier = 5,
+        }
+
+        private enum DocumentsTable
```
👍

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -88,6 +116,11 @@ static DocumentsStorage()
             Slice.From(StorageEnvironment.LabelsContext, "DeletedEtags", ByteStringType.Immutable, out DeletedEtagsSlice);
             Slice.From(StorageEnvironment.LabelsContext, "ConflictedCollection", ByteStringType.Immutable, out ConflictedCollectionSlice);
+            Slice.From(StorageEnvironment.LabelsContext, "AttachmentsKey", ByteStringType.Immutable, out AttachmentsKeySlice);
+            Slice.From(StorageEnvironment.LabelsContext, "AttachmentsKeyIndexSlice", ByteStringType.Immutable, out AttachmentsKeyIndexSlice);
```
Remove `Slice` from the name

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -88,6 +116,11 @@ static DocumentsStorage()
+            Slice.From(StorageEnvironment.LabelsContext, "AttachmentsKeyIndexSlice", ByteStringType.Immutable, out AttachmentsKeyIndexSlice);
+            Slice.From(StorageEnvironment.LabelsContext, "Attachments", ByteStringType.Immutable, out AttachmentsSlice);
+            Slice.From(StorageEnvironment.LabelsContext, "AttachmentsMetadata", ByteStringType.Immutable, out AttachmentsMetadataSlice);
+            Slice.From(StorageEnvironment.LabelsContext, "AttachmentsEtagSlice", ByteStringType.Immutable, out AttachmentsEtagSlice);
```
Remove `Slice` from the name

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -175,27 +203,48 @@ full name
+
+
+            // The attachments schema is as follows
+            // 7 fields (lowered document id, recored separator, lowered name, etag, name, content type, last modified, stream identifier)
```
`recored` is typo

fitzchak replied:
Fixed

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -175,27 +203,48 @@ full name
+            // The attachments schema is as follows
+            // 7 fields (lowered document id, recored separator, lowered name, etag, name, content type, last modified, stream identifier)
+            // We are you using the record separator in order to avoid loading another files that has the same key prefix, 
+            //      e.g. fitz(record-separator)profile.png and fitz0(record-separator)profile.png, without the record separator we would have to load also fitz0 and filter it.
```
Bigger problem would be:

`fitz0(record-separator)profile.png` and `fitz(record-separator)0profile.png` would be indistinguishable

fitzchak replied:
I'm not following. With a separator I can see by prefix: fitz(record-separator) and get all the attachments of it without attachments of fitz0.

ayende replied:
I'm saying it would be a problem _without_ a separator.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                {
+                    throw new InvalidOperationException("Cannot overwrite exisitng attachment.");
+
+                    /*int size;
```
Are we going to support it?

fitzchak replied:
I think so, later on.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                }
+
+                // Insert the stream
+                var tree = context.Transaction.InnerTransaction.CreateTree(AttachmentsSlice);
```
Might be better to drop the storage id notion. The issue is that we are going to have to do accounting, and we want to be able to show the stream name to the user (if an attachment is very large, we want to be able to tell what it is).
Let us use the attachment key as the stream key as well, so we'll be able to handle it.

fitzchak replied:
OK

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                var data = new BlittableJsonReaderObject(tvr.Read((int) DocumentsTable.Data, out size), size, context);
+                var collectionName = ExtractCollectionName(context, documentId, data);
+                var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collectionName.GetTableName(CollectionTableType.Documents));
+                if (table.ReadByKey(loweredKey, out tvr) == false)
```
Shouldn't we do this (and all the work in the method in general) _before_ we run all the attachment stuff?
This is a precondition, after all.

fitzchak replied:
I changed it so we update the document before we save the attachment, but still the document etag will be greater than the attachment.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                var newEtag = GenerateNextEtag();
+                var newEtagBigEndian = Bits.SwapBytes(newEtag);
+
+                var oldChangeVector = tvr.Pointer != null ? GetChangeVectorEntriesFromTableValueReader(ref tvr, 4) : null;
```
How can we have a null change vector here?

fitzchak replied:
We cannot. Fixed.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+        public IEnumerable<LazyStringValue> GetAttachmentNamesForDocument(DocumentsOperationContext context, LazyStringValue lowerDocumentId)
+        {
+            Slice startSlice;
+            // TODO: Should we use here an enumerator or reutrn an array and dispose?
```
Yes, `yield` and dispose please

fitzchak replied:
Done, but it means that we may load the attachments in cases that we are not needing them (if we have a case that load without sending to the user).

ayende replied:
I'm not sure that I'm following here?

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+            keySlice = new Slice(SliceOptions.Key, keyMem);
+
+            return new DisposableAction(() =>
+            {
```
Better use a struct here

fitzchak replied:
Fixed

ayende replied:
No, `IDisposable` is an interface, this is allocting

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+
+            return new DisposableAction(() =>
+            {
+                if (keyMem.HasValue)
```
Better use a struct here

fitzchak replied:
This is true for all Voron.Util.DisposableAction so I changed it to struct.

ayende replied:
You can't. This is allocating because you are returning `IDiposable`.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+            {
+                document = Get(context, lowerDocumentId);
+                if (document == null)
+                    throw new InvalidOperationException($"Cannot delete attachment {name} on a not exist document '{documentId}'.");
```
'not exist' -> 'non existent'
And why is this an error? If they don't have an etag, that is fine and will be a noop

fitzchak replied:
Fixed.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+
+            var file = GetAttachment(context, keySlice);
+            if (file == null)
+                return; //NOP, already deleted
```
No, if I have an etag, should error here.

fitzchak replied:
Done, also for documents delete.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                HttpContext.Response.Headers[Constants.Headers.Etag] = "\"" + attachment.Etag + "\"";
+
+                // Ensure that files are not cached at the browser side.
+                HttpContext.Response.Headers["Cache-Control"] = "no-cache, no-store, must-revalidate";
```
Why is this required? The etag should handle that, no? How is this different from any other endpoint?

fitzchak replied:
OK, I removed it. 
I learned to do this from the code of v3.5.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+
+                HttpContext.Response.Headers["Content-Type"] = attachment.ContentType.ToString();
+
+                var stream = Database.DocumentsStorage.GetAttachmentStream(context, attachment.StreamIdentifier);
```
using

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                var stream = Database.DocumentsStorage.GetAttachmentStream(context, attachment.StreamIdentifier);
+                if (stream == null)
+                {
+                    throw new FileNotFoundException("Attachment is not found. This should not happen.");
```
move to separate method

fitzchak replied:
This was changed already to different impl.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                var name = GetQueryStringValueAndAssertIfSingleAndNotEmpty("name");
+                var contentType = GetStringQueryString("contentType", false) ?? "";
+
+                var tempPath = GetUniqueTempFileName(Path.GetTempPath(), "tmp");
```
Don't use the global temp path, instead, use `DocumentStorage.Environment.Options.DataPager.TempPath`

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                var tempPath = GetUniqueTempFileName(Path.GetTempPath(), "tmp");
+                using (var file = new FileStream(tempPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 4096, FileOptions.DeleteOnClose) )
+                {
+                    await RequestBodyStream().CopyToAsync(file, 81920, Database.DatabaseShutdown);
```
This allows 80 KB, why? You can copy manually using the managed buffer from the context and save all those allocations

fitzchak replied:
Done, with a comment to improve later:

                    using (context.GetManagedBuffer(out buffer))
                    {
                        var requestStream = RequestBodyStream();
                        // TODO: Improve this to be more efficient: read more from the network while we writing to disk.
                        while (await requestStream.ReadAsync(buffer.Buffer.Array, 0, buffer.Length, Database.DatabaseShutdown) > 0)
                        {
                            await file.WriteAsync(buffer.Buffer.Array, 0, buffer.Length, Database.DatabaseShutdown);
                        }

ayende replied:
Not sure if we want it, this is okay, and avoid overloading the system.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                    {
+                        writer.WriteStartObject();
+
```
Add doc id & file name

fitzchak replied:
In the client I'm returning it from the supplied info to the server... But yes, I'll change it to return this data, despite of the duplication.

ayende replied:
I'm thinking about the debuggability of this over the wire, so that would make things easier. Also make sure that we get the proper casing.

fitzchak replied:
I added a todo comment about returning the actual casing on disk as it may not be needed here and will require a big change to the code.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+            }
+        }
+
+        const string TempFilePrefix = "RavenDB.";
```
RavenDB is incredibly generic term for us.  And why do we need / want to have a meaningful name here?
Call it `attachment-{Guid.NewGuid()}` and be done with it.

fitzchak replied:
See current implementation.

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,207 @@
+                var contentType = GetStringQueryString("contentType", false) ?? "";
+
+                var tempPath = GetUniqueTempFileName(Path.GetTempPath(), "tmp");
+                using (var file = new FileStream(tempPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 4096, FileOptions.DeleteOnClose) )
```
Can also specify sequential flag here, probably

fitzchak replied:
OK

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -170,16 +169,63 @@ public unsafe void WriteString(LazyStringValue str)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteStringJoin(IEnumerable<LazyStringValue> strArray)
```
No, why would we ever want to do something like that? You can't undo this to component parts easily.
Why not use an array?

fitzchak replied:
Please review the change I made.

ayende commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -1147,6 +1147,13 @@ private static void WriteMetadata(BlittableJsonTextWriter writer, Document docum
+                {
+                    writer.WriteComma();
+                    writer.WritePropertyName(Constants.Documents.Metadata.Attachments);
+                    writer.WriteStringJoin(document.Attachments);
```
No, use an array here.

fitzchak replied:
ok

ayende replied:
Still need to fix this.

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -170,16 +169,63 @@ public unsafe void WriteString(LazyStringValue str)
             if (numberOfEscapeSequences == 0)
             {
                 WriteRawString(strBuffer, size);
+            }
```
No, this was written intentionally in this way, so there would be no jumps in the code in the common case.

fitzchak replied:
Please review the change I made.

ayende commented on test/FastTests/Client/Attachments/AttachmentsCrud.cs:
```diff
@@ -0,0 +1,70 @@
+                {
+                    "profile.png",
+                    "background-photo.jpg",
+                    //TODO Not working with hebrew id: "fileNANE_#$1^%_בעברית.txt"
```
Add Russian and Thai ids as well, they also gave us issues.

fitzchak replied:
Is this be design or we are going to support it later on?

ayende replied:
Oh, we are going to support it, absolutely.

fitzchak replied:
Yes, this is why I tested against it and I'm leaving this as a comment.

Commit:
RavenDB-6361 Get request should not set the etag manually as the request executer will set it for us

Commit:
RavenDB-6361 Change request.Headers.IfMatch.Add(new EntityTagHeaderValue($"\"{_etag}\""));
to request.Headers.TryAddWithoutValidation("If-Match", $"\"{etag.Value}\"");
This will avoid no needed class allocation and some validation code but I don't see a way to avoid string concat as we need to pass a string there.

Commit:
RavenDB-6361 Stream are not going to be cached so we won't send an etag when requesting an attachment.

Commit:
RavenDB-6361 GetAttachmentCommand should write to a stream that the user supply. This way can dispose the response stream right after the user handle it.

Commit:
RavenDB-6361 minor - fix typo

Commit:
RavenDB-6361 Use the full key name of attachment also for the stream in order to be able to associate each stream with its key easily. It will be needed in troubleshooting big files.

Commit:
RavenDB-6361 There is no reason that we won't have the browser or other client cache the attachment file if he decides so.

Commit:
RavenDB-6361 Update the document before we save the attachment, this way we throw upfront if the document is not exist, but make sure that the document etag will be greater than the attachment.

Commit:
RavenDB-6361 Change vector cannot be null.

Commit:
RavenDB-6361 Read the attachments for the document upfront without waiting for the json reader to read it. This way we can dispose the key faster. But it means that we may load the attachments also when we not using them.

Commit:
RavenDB-6361 DocumentStorage.GetAttachmentKey should be use a DisposableAction which is a struct. This is true also for all other uses of this class as of now.

Commit:
RavenDB-6361 Throw if request to delete a non existent attachment with expected etag. If the document is not exist but we do not expect a specific etag this will be a no-op.

Commit:
RavenDB-6361 Use storage temp folder instead of system one.
Format the temp file name as: attachment.123456ABC.put
Also use SequentialScan

Commit:
RavenDB-6361 Better implementation of copy the stream to file. Still need to work on it.
Return the attachment name, document id and content type when calling PutAttachment despite that the user just sent this data in the request.

Commit:
RavenDB-6361 Fix NRE and refactor code

Commit:
RavenDB-6361 Make the code more efficient for the common use case

Commit:
RavenDB-6361 Add a comment about returning the proper casing of document id.

fitzchak commented:
WIP - but can be reviewed.
Known issue: the test is failing sometimes with empty attachments... I guess there is an issue with memory management. Will investigate later.

Commit:
RavenDB-6361 Better error message

Commit:
RavenDB-6361 Fix merge error

ayende commented on src/Raven.Client/Documents/Operations/GetAttachmentOperation.cs:
```diff
@@ -0,0 +1,85 @@
+            private readonly string _name;
+            private readonly Action<Stream> _handleStreamResponse;
+
+            public GetAttachmentCommand(string documentId, string name, Action<Stream> handleStreamResponse)
```
The action should also get the `AttachmentResult`, so user can just provide generic method to handle anything.

fitzchak replied:
I'm not sure that I'm following. We do return here the AttachmentResult. Do you want it to be passed in the constructor? The content type is not needed there.

ayende commented on src/Raven.Client/Http/RavenCommand.cs:
```diff
@@ -88,11 +102,28 @@ protected static void ThrowInvalidResponse()
+
+        protected void AddEtagIfNotNull(long? etag, HttpRequestMessage request)
+        {
+#if DEBUG
```
👍

ayende commented on src/Raven.Server/Documents/Attachment.cs:
```diff
@@ -0,0 +1,38 @@
+        public LazyStringValue ContentType;
+        public Stream Stream;
+
+        public static string Canonize(string name, bool trimEnd = true)
```
No one is calling this method

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -176,27 +203,48 @@ full name
+
+
+            // The attachments schema is as follows
+            // 5 fields (lowered document id AND record separator AND lowered name, etag, name, content type, last modified, stream identifier)
```
Didn't we remove the stream identifier?

fitzchak replied:
Yes, we did. Fixed the comment.

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -176,27 +203,48 @@ full name
+                Count = 1,
+                Name = AttachmentsKeySlice
+            });
+            AttachmentsSchema.DefineIndex(new TableSchema.SchemaIndexDef
```
We can remove it and just provide the same operations on the key

fitzchak replied:
We should support table.SeekForwardFrom(AttachmentsSchema.Key, startSlice, true) but it is failing now because of expecting fixedSizeTree

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,220 @@
+                HttpContext.Response.Headers["Content-Disposition"] = $"attachment; filename=\"{fileName}\"";
+                HttpContext.Response.Headers["Content-Type"] = attachment.ContentType.ToString();
+                HttpContext.Response.Headers[Constants.Headers.Etag] = "\"" + attachment.Etag + "\"";
+                // TODO: HttpContext.Response.Headers["Content-Range"] = ;
```
Don't think we want to support this

fitzchak replied:
OK

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,220 @@
+
+                using (var stream = attachment.Stream)
+                {
+                    await stream.CopyToAsync(ResponseBodyStream());
```
No, copy this using the managed buffer on the context

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,220 @@
+
+                AttachmentResult result;
+                var tempPath = GetUniqueTempFileName(Database.DocumentsStorage.Environment.Options.DataPager.Options.TempPath, "attachment.", "put");
+                using (var file = new FileStream(tempPath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, 4096, FileOptions.DeleteOnClose | FileOptions.SequentialScan))
```
On Linux we'll need to delete this manually, need to handle this too

fitzchak replied:
Done

ayende commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,220 @@
+        private static string GetUniqueTempFileName(string tempPath, string prefix, string extension)
+        {
+            int attempt = 0;
+            while (true)
```
Why not use a Guid here, then you can be sure it won't have conflicts

fitzchak replied:
OK

ayende commented on src/Sparrow/Json/BlittableJsonTextWriter.cs:
```diff
@@ -173,13 +172,59 @@ public unsafe void WriteString(LazyStringValue str)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void WriteStringJoin(LazyStringValue[] strArray)
```
No, this method needs to be deleted. You need to use a json array instead. We don't do string parsing.

fitzchak replied:
In the metadata we support only strings values for now. 
We can either change this requirement or use string.Join and not use LazyStringValue for @attachments metadata.

ayende commented on src/Voron/Util/DisposableAction.cs:
```diff
@@ -9,9 +9,9 @@
+    /// A helper struct that translate between Disposable and Action
     /// </summary>
-    public class DisposableAction : IDisposable
+    public struct DisposableAction : IDisposable
```
No, bad change. Instead, we need a _distinct_ struct whenever this is used.

Commit:
RavenDB-6361 After deleting the attachment we now increase the etag of the document that holding this attachment and removing the HasAttachment flag if there are no more attachments.

Commit:
RavenDB-6361 Fix comment

Commit:
RavenDB-6361 Fix requested changes after PR review

Commit:
RavenDB-6361 Implement DeleteAttachmentCommand in the client side and add test for delete.
Use SeekByPrimaryKey and fix a bug in OpenTable method with a slice argument.
Fix bugs.

Commit:
RavenDB-6361 Lazy load the attachments only if we send it to the user

Commit:
RavenDB-6361 We do not delete the attachment's metadata but instead delete the stream only and update the metadata with a delete marker

fitzchak commented:
Ready to merge

ppekrol commented on src/Raven.Server/Documents/Handlers/AttachmentHandler.cs:
```diff
@@ -0,0 +1,209 @@
+                var contentType = GetStringQueryString("contentType", false) ?? "";
+
+                AttachmentResult result;
+                var tempPath = Path.Combine(Database.DocumentsStorage.Environment.Options.DataPager.Options.TempPath, $"attachment.{Guid.NewGuid():N}.put");
```
Use Database.Configuration.Storage.TempPath that is a PathSetting and contains Combine that takes care of all platform related path issues?

ppekrol commented:
@fitzchak 1000+ tests are failing here, can you look into that?

ayende commented on src/Raven.Client/Http/RequestExecuter.cs:
```diff
@@ -249,8 +249,9 @@ private static HttpRequestMessage CreateRequest<TResult>(ServerNode node, RavenC
-                    else
+                    else if (command.ResponseType == RavenCommandResponseType.Array)
                         command.SetResponse((BlittableJsonReaderArray)null, fromCache: false);
+                    command.SetResponse(null, null, 0, fromCache: false);
```
This looks like a missing else here.

ayende commented on src/Raven.Client/Json/MetadataAsDictionary.cs:
```diff
@@ -22,10 +23,20 @@ public void Init()
+            var arr = value as BlittableJsonReaderArray;
+            if (arr != null)
+            {
+                return string.Join(",", arr.Items.Select(x => x.ToString()));
```
Better to avoid using delegate & linq like that here.

ayende replied:
We need to consider allowing more complex types in the metadata ( I really don't want to do that, though ).
Maybe just for our stuff?

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2596,5 +2631,375 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+        }
+
+        private bool ThrowDocumentNotFound(DocumentsOperationContext context, Slice loweredKey, string documentId, string name, bool isDelete, long? expectedEtag)
+        {
```
Too big and does too much, and the name is misleading

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2596,5 +2631,375 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                        transactionMarker
+                    };
+
+                    table.Update(tvr.Id, tbv);
```
This is actually invalid for us to do, we have to copy it to the size and then update.
@karmeli87 has run into that, and it can cause corruption in strange cases.

karmeli87 replied:
In this case I think it safe, because every call of this function "data" and "LoweredKey" are a new objects (btw, do you need to dispose the data object?).
So we can safely update the table without corruption.

ayende replied:
It isn't actually new objects, they point to the old data. 
This is actually safe because we don't change the size of the data, but that isn't something you can count on (it may be a value that came from replication, then you add your change vector to it) and then it would only fail on very rare cases, but that is data corruption.

ayende commented:
Closing this because I rebased and squashed everything
